server/pxf-hdfs/src/main/java/org/greenplum/pxf/plugins/hdfs/parquet/ParquetUtilities.java

```diff
@@ -0,0 +1,119 @@
+            }
+        }
+        return data;
+    }
+
+    /**
+     * Decode a String representation of a Greenplum object into a Java object according to primitiveTypeName
+     *
+     * @param val                   A String representation of a Greenplum object
+     * @param primitiveTypeName     Primitive type name of this object
+     * @param logicalTypeAnnotation Logical type annotation of this object if it has one
+     * @return A Java object
+     */
+    private Object decodeString(String val, PrimitiveType.PrimitiveTypeName primitiveTypeName, LogicalTypeAnnotation logicalTypeAnnotation) {
+        if (val == null) {
+            return null;
+        }
+        switch (primitiveTypeName) {
+            case BINARY:
+                if (logicalTypeAnnotation instanceof LogicalTypeAnnotation.StringLogicalTypeAnnotation) {
+                    return val;
+                } else {
+                    return pgUtilities.parseByteaLiteral(val);
+                }
+            case BOOLEAN:
+                //parquet bool val is "true" or "false" but pgUtilities only accept "t" or "f"
+                return pgUtilities.parseBoolLiteral(val);
+            case INT32:
+                if (logicalTypeAnnotation instanceof LogicalTypeAnnotation.DateLogicalTypeAnnotation) {
+                    // ParquetResolver.fillGroupWithPrimitive will convert Date String into INT32 (number of days from the Unix epoch, 1 January 1970)
+                    return val;
+                } else if (logicalTypeAnnotation instanceof LogicalTypeAnnotation.IntLogicalTypeAnnotation &&
+                        ((LogicalTypeAnnotation.IntLogicalTypeAnnotation) logicalTypeAnnotation).getBitWidth() == 16) {
+                    return Short.parseShort(val);
+                } else {
+                    return Integer.parseInt(val);
+                }
+            case INT64:
+                return Long.parseLong(val);
+            case DOUBLE:
+                return Double.parseDouble(val);
+            case FLOAT:
+                return Float.parseFloat(val);
+            case INT96:
+            case FIXED_LEN_BYTE_ARRAY:
+                return val;
+            default:
+                throw new PxfRuntimeException(String.format("type: %s is not supported", primitiveTypeName));
+        }
+    }
+
+    /**
+     * Provide more detailed error message
+     *
+     * @param isMultiDimensional Whether this error is caused by inserting a multi-dimensional array
+     * @param val                The string representation we are currently trying to decode
+     * @return The generated error message
+     */
+    private String createErrorHintFromValue(boolean isMultiDimensional, String val) {
+        if (isMultiDimensional) {
+            return String.format("Column value \"%s\" is a multi-dimensional array, PXF does not support multi-dimensional arrays for writing Parquet files.", val);
```

Comment:
Style-nit:

```suggestion
            return String.format("Column value \"%s\" is a multi-dimensional array, PXF does not support multi-dimensional arrays when writing Parquet files.", val);
```

This is a weak suggestion because the message as it is now matches with the message in `OrcUtilities` but the phrase `for writing Parquet files` sounds awkward to me. Maybe a better error message would be

```suggestion
            return String.format("Column value \"%s\" is a multi-dimensional array; PXF does not support writing multi-dimensional arrays to Parquet files.", val);
```